### PR TITLE
Jetpack UX: Add xml and rest api error handling

### PR DESCRIPTION
**Pull request by kozer**
Related to https://github.com/Automattic/dotcom-forge/issues/3241

## Proposed Changes

This pr add support for jetpack XML-RPC and REST API connection is

### Test 1: REST API errors
1. Apply D119857-code
2. Create a new WoA site.
3. Enable ssh.
4. Visit eg `/posts` page
5. Login to the site and add the following to your `wp-config`
```
if ( ! empty( $_GET['rest_route'] ) ) {
	http_response_code( 404 );
	die('not found');
}
```
6. Ensure that you see the following error:
rest_api_jetpack

### Test 2: XML-RPC errors
1. Apply D119857-code
2. Create a new WoA site.
3. Enable ssh.
4. Visit eg `/posts` page
5. Login to the site and add the following to your `wp-config`
```
if ( ! empty( $_GET['for'] ) && $_GET['for'] === 'jetpack' ) {
	http_response_code( 404 );
	die('not found');
}
```
6. Ensure that you see the following error:
xml-rpc_jetpack

## Pre-merge Checklist

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?

### CODE REVIEW

**Review by danielbachhuber — COMMENTED** [review]
Can you include test instructions? Also, should we link to the new support doc in these messages?

File: client/components/jetpack/connection-health/index.tsx [hunk 1]
```diff
@@ -79,6 +81,34 @@ export const JetpackConnectionHealthBanner = ( { siteId }: Props ) => {
 		);
 	}
 
+	if ( errorType === REST_API_ERROR ) {
+		return (
+			<ErrorNotice
+				errorType={ errorType }
+				errorText={ translate( 'Jetpack encounters REST API connection issues.' ) }
+				noticeActionHref={ localizeUrl(
+					'https://wordpress.com/support/why-is-my-site-down/#theres-an-issue-with-your-sites-jetpack-connection'
```
Comment (danielbachhuber) [inline client/components/jetpack/connection-health/index.tsx]: I created an issue to add an explanation to Resolve Jetpack Errors: https://github.com/Automattic/en.support-docs-content/issues/2080

Comment (kozer) [reply client/components/jetpack/connection-health/index.tsx]: Should I create a ticket to revisit this and update the URL after the new docs are in place?

**Comment by danielbachhuber** [comment]
@sejas Could you give this a review tomorrow?

File: client/state/jetpack-connection-health/selectors/is-jetpack-connection-problem.js [hunk 2]
```diff
@@ -19,3 +23,53 @@ export default function isJetpackConnectionProblem( state, siteId ) {
...
+/**
+ * React HOC to check if the current site has possible Jetpack connection problem.
+ */
+export const withJetpackConnectionProblem = ( Wrapped ) => ( props ) => {
+	const siteId = useSelector( ( state ) => getSelectedSiteId( state ) );
+	const isPossibleConnectionProblem = useIsJetpackConnectionProblem( siteId );
+	return (
+		<Wrapped { ...props } isPossibleJetpackConnectionProblem={ isPossibleConnectionProblem } />
+	);
+};
```
Comment (sejas) [inline client/state/jetpack-connection-health/selectors/is-jetpack-connection-problem.js:80]: Nice HOC.

File: client/state/jetpack-connection-health/selectors/is-jetpack-connection-problem.js [hunk 3]
```diff
@@ -19,3 +23,53 @@ export default function isJetpackConnectionProblem( state, siteId ) {
...
+		 * This event listener is used to detect if the site is inaccessible.
+		 * Among others it's used as a login in `wpcom-proxy-request` package,
+		 * and the idea is borrowed from there to detect the site inaccessibility.
+		 */
+		window.addEventListener( 'message', onMessage );
+
+		return () => {
+			window.removeEventListener( 'message', onMessage );
+		};
+	}, [ isPossibleConnectionProblem, siteId, dispatch ] );
```
Comment (sejas) [inline client/state/jetpack-connection-health/selectors/is-jetpack-connection-problem.js]: I was trying to find other alternatives to this global listener, but I cannot imagine another way to do it.

**Review by sejas — APPROVED** [review]
I tested this PR along with D119857-code and it works as described.

The only two things I noticed that are not related to this PR are:
1. The jetpack error is cached and keeps showing me the error even after I've fixed it.
  - Between tests I had to change the health Jeptack cahce key on WPcom.
2. Sometimes when accessing to the site url I'm redirected to the global version
   - Example: `http://calypso.localhost:3000/posts/20230830testingjetpackresterror.wordpress.com` -> `http://calypso.localhost:3000/posts`

## Screenshots:

Before this PR:
Screenshot 2023-08-30 at 14 07 11



After this PR REST:

Screenshot 2023-08-30 at 14 13 27

After this PR XML-RPC:

Screenshot 2023-08-30 at 14 08 59

File: client/state/jetpack-connection-health/selectors/is-jetpack-connection-problem.js (hunk 3 above)
Comment (kozer) [reply client/state/jetpack-connection-health/selectors/is-jetpack-connection-problem.js]: Yes, I searched all Jetpack reducer stuff before I ended up with that solution. I ended up enabling debug and noticed that the only consistent thing was the `wpcom-proxy-request` package. 
Searching a bit realized that the messages I saw were introduced using the same way I followed here.

**Comment by kozer** [comment]
> The only two things I noticed that are not related to this PR are:

> The jetpack error is cached and keeps showing me the error even after I've fixed it.
> Between tests I had to change the health Jeptack cache key on WPcom.

Sorry, I had to mention this in the description of the PR. We cache the request from day one we released this new feature. I see how this can become a problem, however, it was added for performant reasons ( Testing Jetpack connection is a bit expensive )

> Sometimes when accessing to the site URL I'm redirected to the global version

This is something that is happening to me a lot lately, and I thought it was something wrong with the local environment or something. For me, this is happening on refresh. Does the same apply to you? Also, does this happen to you all the time, or when Jetpack has an error? Maybe it is worth to create a new ticket and investigate more on that, especially if this is happening on production.

**Comment by danielbachhuber** [comment]
> > Sometimes when accessing to the site URL I'm redirected to the global version
> 
> This is something that is happening to me a lot lately, and I thought it was something wrong with the local environment or something.

The Redux cache is purged randomly when developing locally.

**Comment by a8ci18n** [comment]
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/8829018

Thank you @kozer for including a screenshot in the description! This is really helpful for our translators.

**Comment by a8ci18n** [comment]
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/8829018).

**Pull request merged**